### PR TITLE
Ood support

### PR DESCRIPTION
**Pull request by osihlovec**
Modifies the ruleset.py and seq_generator.py files, such that they generate o.o.d. data in addition to regular train/test data. Additionally, the generator ensures that it only utilizes production rules seen during training when sampling the validation data. We should probably store each category separately (training, ood validation, regular validation), but in the very least, ood should be separate (such that these records are not used to train the model). If you have any suggestions, then please share them via response to this PR or WhatsApp.

UPDATE: I have noticed that there is an issue with overlap between training and test_id sets. I have now corrected the mistake, but I have noticed that the size of the resulting dataset need not be equal to the loop iterations, since the sampled data point might already be contained in the set, in spite of incrementing the iterator.

### CODE REVIEW

File: seq_generator.py [hunk 1]
```diff
@@ -84,13 +87,26 @@ def inc_repeats():
...
     with open(loc_dataset, "w") as f:
-        json.dump(r, f, indent=4)
+        json.dump(train_data, f, indent=4)
+
+    non_terminals = list(grammar.rules.keys())
+
+    # omit unproductive rules from test sets
+    for nt in non_terminals:
+        for rule in grammar.rules[nt]:
+            if (not rule[1]):
```
Comment (jnice-81) [inline seq_generator.py:117]: I am not 100 % sure if this is correct. You mark a rule (e.g. rule 3) as used if it ever appeared during generation right? However the way we describe our dataset rule 3 might be able to produce [4, 5], "a" or [1, 2]. If now rule 3 appears during generation and the rule 3 -> [4, 5] is used then the way this is written now would simply say that rule 3 has been used, but actually the rule 3 -> "a" or 3 -> [1, 2] might have never been used.

File: seq_generator.py [hunk 2]
```diff
@@ -84,13 +87,26 @@ def inc_repeats():
...
+                grammar.rules[nt].remove(rule)
+    
+    # not sure how to export/save the data yet, will be completed later
```
Comment (jnice-81) [inline seq_generator.py:122]: You can use the generate_data function (also in seq_generator). It is on the main branch.

File: seq_generator.py (hunk 1 above)
Comment (osihlovec) [reply seq_generator.py:117]: No, the way I mark rules is that I extend every "subrule" into a list whose second entry corresponds to the used flag. In this stage, I delete every subrule whose second entry is False. Thank You for making sure though.
image

File: seq_generator.py (hunk 2 above)
Comment (osihlovec) [reply seq_generator.py:122]: I apologize but I am not quite sure what you mean. The create_dataset function which I modified still makes calls to the existing generate_data method. Could you please elaborate a bit further?

File: seq_generator.py (hunk 1 above)
Comment (osihlovec) [reply seq_generator.py:117]: Maybe I can rename the variable such that this is clearer?

Comment (jnice-81) [reply seq_generator.py:117]: Ahh I see. Naa guess this is fine then

File: seq_generator.py (hunk 2 above)
Comment (jnice-81) [reply seq_generator.py:122]: Sorry I confused myself when reading the code. What I meant is that you can use the same approach to export the data as the create_data function on main does. (Basically just dump whatever generate_data returns as json)

**Pull request merged**